### PR TITLE
CLN: use f-strings where possible

### PR DESCRIPTION
**Pull request by akx**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
  - No new tests, behavior hasn't changed (except possibly become a bit faster)
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
  - Nothing new.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.
  - No bug or feature, just cleaning.

---

This PR replaces a bunch of string concatenations and other non-modern formatting operations with f-strings. 

The first commit was done mechanically with [`flynt`](https://github.com/ikamensh/flynt), followed up by further manual touch-ups that `flynt` wasn't smart enough to touch.

This does _not_ touch a bunch of concatenations in `pandas/io/stata.py`, since those are covered by cln commits in #49228.

### CODE REVIEW

File: pandas/_version.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ def run_command(commands, args, cwd=None, verbose=False, hide_stderr=False, env=
             if e.errno == errno.ENOENT:
                 continue
             if verbose:
-                print("unable to run %s" % dispcmd)
```
Comment (jbrockmendel) [inline pandas/_version.py:91]: this file is auto-generated, so we dont edit it directly

File: pandas/core/arrays/string_arrow.py [hunk 2]
```diff
@@ -345,14 +345,14 @@ def _str_match(
         self, pat: str, case: bool = True, flags: int = 0, na: Scalar | None = None
     ):
         if not pat.startswith("^"):
-            pat = "^" + pat
```
Comment (jbrockmendel) [inline pandas/core/arrays/string_arrow.py:348]: i dont think these are an improvement.  stick to places that currently use .format

Comment (akx) [reply pandas/core/arrays/string_arrow.py:348]: Would this apply to all string concatenation? I think e.g. the changes in `sql.py` (https://github.com/pandas-dev/pandas/pull/49229/files#diff-9268174bfb15f08ef2267375665a85fecf201999902542f6fc9c0d3fadfb4553 if GitHub feels like linking correctly) read much better as an f-string, for one?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Thanks for your PR

There's some failing tests, if it's not an automated check that gets it right the first time (rather than one that requires manual edits) then I think we should pass on this one, 120 files are too many to review manually

I remember reading that flynt's rewrites were sometimes unsafe, let's not risk it

**Comment by akx** [comment]
Hi @MarcoGorelli, thanks for the review. I remade this PR to limit the scope to non-test files – the test failure was caused by `flynt`'s aggressive string concatenation mode changing things it shouldn't have there.

In addition, I took extra care to only mechanically run safe transforms, and then applied other select fixes by hand in the subsequent commits.

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
I'm all for automation, but if a tool is too aggressive and requires manual fixups, that's a bit of a red flag and I don't think we should be using it, sorry. This is still failing various tests

I'd suggest looking at https://github.com/pandas-dev/pandas/issues/48855 if you're interested in linting issues

**Comment by MarcoGorelli** [comment]
Also, we already have pyupgrade for rewriting f-strings when it's safe to do so

Closing for now then, but thanks for your PR

**Comment by akx** [comment]
@MarcoGorelli I would ask you to reconsider. This PR isn't really done by an automated tool anymore. I used `flynt` as a guide, and made the changes by hand.

**Comment by MarcoGorelli** [comment]
The issue with adding mostly cosmetic changes without an automated tool is that there's no way to ensure contributors won't new code that goes against these changes, and then we'll be here again in a month's time making the same changes manually

But if there's a case where this makes a performance difference at the macro level, then happy to take that- e.g. the change you mentioned here https://github.com/pandas-dev/pandas/pull/49229#discussion_r100247531

**Comment by akx** [comment]
> The issue with adding mostly cosmetic changes

These aren't really only cosmetic changes.

> without an automated tool is that there's no way to ensure contributors won't new code that goes against these changes

As far as I understand, all PRs to Pandas are reviewed by humans too.  
That reviewing human would probably go "Hey, couldn't you use `f"^{foo}"` instead of `"^" + str(foo)`" (or similar).

Again, these changes have been made by a human (yours truly), not an automated tool. Would the context of the review be different if I had not originally mentioned a tool?

**Comment by MarcoGorelli** [comment]
These kinds of changes should be made by automated tools without requiring manual fixups, so that
- we can add this to CI
- we don't need to keep leaving comments in PR reviews about them

If it's done manually, then we'll be doing it manually again in a month's time, and then again, and so on

**Comment by akx** [comment]
I'm sorry, but can you explain what "these kinds of changes" are? 

These changes hadn't, and could have, been made automatically by a tool in CI because no tool currently exists that could do the semantic/dataflow analysis to see whether a string concatenation can be safely replaced with an f-string.

Fortunately, humans _can_ do that, and as said in my comment after the first test failures, I manually, by hand, vetted each change recommended by `flynt`.

**Comment by akx** [comment]
Ah, darnit, apparently trying to rebase/force-push a closed PR destroys things and this can't be reopened anymore.

Considering the discussion above, would you be open to re-reviewing a PR if I open a new one, @MarcoGorelli? I honestly believe these changes are objectively _good_ and could have been done earlier by humans, but it just hadn't happened.

**Comment by MarcoGorelli** [comment]
I just meant mostly cosmetic / stylistic changes - if any of these delivers a measurable performance improvement, then sure, feel free to submit a PR

If there's no performance improvement and it can't be rewritten/checked automatically, then I'm not sure there's much reason to make the change

**Comment by akx** [comment]
@MarcoGorelli Would https://twitter.com/raymondh/status/1205969258800275456 (found via the https://pylint.pycqa.org/en/latest/user_guide/messages/convention/consider-using-f-string.html doc page) be enough proof that f-strings will deliver performance improvements even in the simple cases?

**Comment by MarcoGorelli** [comment]
that tweet shows savings of a few nanoseconds -if there's some pandas operation where concatenating strings compounds and there's a noticeable performance improvement overall, then that's great - your sql.py example might be a good example of this. if it's just an error message that'd only be printed once, then I'd say this isn't worth the churn

**Comment by akx** [comment]
Well, to put the tweet in other terms, it shows improvements of 32% to 300% for a simple operation.

What's the downside of "churn" for an improvement and how do Pandas developers quantify when it's worth it or not? 

Also, I'd like to point out that this isn't the first PR to just change things to f-strings because it's possible; has something changed since https://github.com/pandas-dev/pandas/issues/29547?

**Pull request closed without merging**